app/src/main/java/fi/helsinki/opintoni/task/CourseImplementationCacheUpdater.java

```diff
@@ -0,0 +1,118 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with MystudiesMyteaching application.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package fi.helsinki.opintoni.task;
+
+import fi.helsinki.opintoni.domain.CachedItemUpdatesCheck;
+import fi.helsinki.opintoni.integration.coursepage.CoursePageClient;
+import fi.helsinki.opintoni.repository.CachedItemUpdatesCheckRepository;
+import fi.helsinki.opintoni.service.TimeService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.cache.Cache;
+import org.springframework.cache.CacheManager;
+import org.springframework.stereotype.Component;
+import org.springframework.web.client.RestClientException;
+
+import java.lang.management.ManagementFactory;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Locale;
+
+import static fi.helsinki.opintoni.cache.CacheConstants.COURSE_PAGE;
+import static java.util.Arrays.asList;
+
+@Component
+public class CourseImplementationCacheUpdater {
+
+    @Value("${locale.available}")
+    private String[] availableLocales;
+
+    private final CoursePageClient coursePageClient;
+    private final CachedItemUpdatesCheckRepository cachedItemUpdatesCheckRepository;
+    private final CacheManager persistentCacheManager;
+
+    private static final Logger log = LoggerFactory.getLogger(CourseImplementationCacheUpdater.class);
+
+    @Autowired
+    public CourseImplementationCacheUpdater(CoursePageClient coursePageClient,
+                                            CachedItemUpdatesCheckRepository cachedItemUpdatesCheckRepository,
+                                            CacheManager persistentCacheManager) {
+        this.coursePageClient = coursePageClient;
+        this.cachedItemUpdatesCheckRepository = cachedItemUpdatesCheckRepository;
+        this.persistentCacheManager = persistentCacheManager;
+    }
+
+    public void getChancedCourseImplementationsAndUpdateCached() {
+        CachedItemUpdatesCheck cachedItemUpdatesCheck = cachedItemUpdatesCheckRepository.findByCacheName(COURSE_PAGE)
+            .orElseGet(this::initialCourseImplementationUpdatesCheck);
+
+        log.info("checking for course implementation updates since {}", cachedItemUpdatesCheck.dateTime);
+
+        try {
+            LocalDateTime updateCheckDateTime = LocalDateTime.now();
+
+            List<Long> updatedCourses =
+                coursePageClient.getUpdatedCourseImplementationIds(
+                    cachedItemUpdatesCheck.dateTime.atZone(TimeService.HELSINKI_ZONE_ID).toEpochSecond());
+
+            updateCachedCourses(updatedCourses);
+
+            cachedItemUpdatesCheck.dateTime = updateCheckDateTime;
```

Comment:
Eikö juuri turvallisinta ole luoda aikaleima (jota käytetään muutosten kyselyn since date -arvona), ennen kuin kutsu lähetetään? Silloin ei varmasti jää pientäkään aikaikkunaa, jolle ei kysellä muutoksia.

Reply:
Ihan oikeassa olet -- mä ajattelin tän jotenkin nurinkurisesti aiemmin. Eihän tuosta tosiaan haittaa ole vaikka aikaleima on "llian kaukana" menneisyydessä muuten kuin siltä osin että haetaan mahdollisesti uudestaan muutoksia mitkä on haettu jo aiemminkin -> nevermind.